File: my-folder/0496-next-greater-element-i/solution.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # def find_next_element(hm,val):
        #     for key,value in hm.items():
        #         if value==val:
        #             return key
        #     return 0 
        if len(nums2)==1:
            return [-1]
        res=[-1]*len(nums1)
        hm={val:i for i,val in enumerate(nums2)}
        for i,num in enumerate(nums1):
            j=hm[num]+1
            # if(need+1>=len(nums2)):
            #     res[i]=-1
            # nextelement = find_next_element(hm,need)
            
            while(j<len(nums2)):
                if num<nums2[j]:
                    res[i]=nums2[j]
                    break
                j+=1
        return res
        
# [4,1,2], [1,3,4,2]  need=2 2+1<=4
#  i


        

File: my-folder/0496-next-greater-element-i/solution.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int length = nums2.length - 1;
        // int[] res = new int[length];
        Map<Integer, Integer> res = new HashMap<>();
        Stack<Integer> stack = new Stack<>();
        for(int i = length; i >= 0; i--){
            int curr = nums2[i];

            while (!stack.isEmpty() && stack.peek() <= curr) {
                stack.pop();
            }
            if(stack.isEmpty()){
                res.put(curr, -1);
            }
            else{
                res.put(curr, stack.peek());
            }

            stack.push(nums2[i]); 
        } 
       

       int[] result = new int[nums1.length];
       for(int i = 0; i < nums1.length; i++){
        result[i] = res.get(nums1[i]);
       }

       return result;
    }
}
